eggplant/market/models/cart.py

```diff
@@ -53,35 +60,49 @@ def add_to_items(self, product=None, quantity=1, delivery_date=None):
 
     def remove_from_items(self, product=None, quantity=1, delivery_date=None):
         current = self.items.filter(product=product,
-                                    delivery_date=delivery_date)
-        if current.exists():
-            if current[0].quantity - quantity > 1:
-                current.update(quantity=models.F('quantity') - quantity)
+                                    delivery_date=delivery_date).first()
+        if current:
+            if current.quantity - quantity > 1:
+                current.quantity -= quantity
+                current.save()
             else:
                 self.items.filter(basket=self, product=product, delivery_date=delivery_date).delete()
 
     def get_total_amount(self):
-        # TODO: This should return a Money type instead of decimal
-        # TODO: When calculating this, we should fail when currencies differ
-        total = Decimal('0')
-        for item in self.items.all():
-            total += item.quantity * item.product.price.amount
-        return total
+
+        # FIXME this check should be removed, it is better to check for the currenices when an item is added
+        # sanity check for the currencies
+
+        if 1 != self.items.all().values('product__price_currency').annotate(total=Count('product__price_currency')).count():
+            raise ValidationError('The products have different currencies')
+
+        total = self.items.all().aggregate(total=Sum(F('product__price') * F('quantity')), cur=Max('product__price_currency'))
+        return Money(total['total'], total['cur'])
```

Comment:
Strange, I tested it on sqlite and it worked. It looks like a difference between psql and sqlite in types. I will rewrite it as a loop. With pre-fetching and selecting related data the performance would be almost the same, more readable and work on every database engine. 

Well it works for me, and it generates a correct query: 
SELECT MAX("market_product"."price_currency") AS "cur", CAST(SUM(("market_product"."price" * "market_basketitem"."quantity")) AS NUMERIC) AS "total" FROM "market_basketitem" INNER JOIN "market_product" ON ("market_basketitem"."product_id" = "market_product"."id") WHERE "market_basketitem"."basket_id" = '19

And I use python 3.4.5